### PR TITLE
switch template controller to external

### PR DESCRIPTION
**Pull request by mfojtik**
Switches the template controller to external types and external build client. This is on a path for eliminating the internal build client entirely.

/cc @bparees 
/cc @deads2k

### CODE REVIEW

File: pkg/template/controller/readiness.go [hunk 1]
```diff
@@ -1,58 +1,65 @@
...
-	buildapi "github.com/openshift/origin/pkg/build/apis/build"
-	buildinternalhelpers "github.com/openshift/origin/pkg/build/apis/build/internal_helpers"
-	buildclient "github.com/openshift/origin/pkg/build/generated/internalclientset"
+	appsutil "github.com/openshift/origin/pkg/apps/util"
 	buildutil "github.com/openshift/origin/pkg/build/util"
-	routeapi "github.com/openshift/origin/pkg/route/apis/route"
 )
 
+// readinessScheme defines a scheme which include only resources this controller understand how to get the readiness checks.
+var readinessScheme = runtime.NewScheme()
```
Comment (mfojtik) [inline pkg/template/controller/readiness.go]: @deads2k scope scheme now

File: pkg/template/controller/templateinstance_controller.go [hunk 2]
```diff
@@ -28,16 +21,33 @@ import (
...
+	templatelister "github.com/openshift/client-go/template/listers/template/v1"
+	"github.com/openshift/origin/pkg/authorization/util"
+	"github.com/openshift/origin/pkg/client/templateprocessing"
 )
 
-const readinessTimeout = time.Hour
+const (
+	readinessTimeout = time.Hour
+
+	WaitForReadyAnnotation    = "template.alpha.openshift.io/wait-for-ready"
```
Comment (mfojtik) [inline pkg/template/controller/templateinstance_controller.go]: @deads2k candidates for openshift/api (will send a follow up with bump)

File: pkg/template/controller/readiness.go [hunk 3]
```diff
@@ -62,105 +69,96 @@ func checkBuildConfigReadiness(oc buildclient.Interface, obj runtime.Object) (bo
...
 // checkJobReadiness determins if a Job is ready, failed or neither.
 func checkJobReadiness(obj runtime.Object) (bool, bool, error) {
-	job := obj.(*batch.Job)
-
+	job := obj.(*batchv1.Job)
 	ready := job.Status.CompletionTime != nil
-	failed := job.Status.Failed > 0
-
-	return ready, failed, nil
+	return ready, job.Status.Failed > 0, nil
```
[the remark this answers is not in the document]
Comment (mfojtik) [reply pkg/template/controller/readiness.go]: yeah, just pure laziness ;-) i found 'failed' to be just always 'false' for some checks, noticed this when I was scraping them

[the remark this answers is not in the document]
Comment (mfojtik) [reply pkg/template/controller/readiness.go]: fixed.

File: pkg/template/controller/readiness.go [hunk 4]
```diff
@@ -62,105 +75,109 @@ func checkBuildConfigReadiness(oc buildclient.Interface, obj runtime.Object) (bo
 // checkDeploymentReadiness determins if a Deployment is ready, failed or
 // neither.
 func checkDeploymentReadiness(obj runtime.Object) (bool, bool, error) {
-	d := obj.(*extensions.Deployment)
+	d, ok := obj.(*kappsv1.Deployment)
```
Comment (deads2k) [inline pkg/template/controller/readiness.go]: You'll need all four flavors of this method for each version.

File: pkg/template/controller/readiness.go [hunk 5]
```diff
@@ -62,105 +75,109 @@ func checkBuildConfigReadiness(oc buildclient.Interface, obj runtime.Object) (bo
...
 
 // checkJobReadiness determins if a Job is ready, failed or neither.
 func checkJobReadiness(obj runtime.Object) (bool, bool, error) {
-	job := obj.(*batch.Job)
-
-	ready := job.Status.CompletionTime != nil
-	failed := job.Status.Failed > 0
-
-	return ready, failed, nil
+	job, ok := obj.(*batchv1.Job)
```
Comment (deads2k) [inline pkg/template/controller/readiness.go]: you'll need all versions of jobs

File: pkg/template/controller/readiness.go [hunk 6]
```diff
@@ -62,105 +75,109 @@ func checkBuildConfigReadiness(oc buildclient.Interface, obj runtime.Object) (bo
...
 func checkStatefulSetReadiness(obj runtime.Object) (bool, bool, error) {
-	ss := obj.(*apps.StatefulSet)
-
-	ready := ss.Status.ObservedGeneration != nil &&
-		*ss.Status.ObservedGeneration == ss.Generation &&
-		ss.Status.ReadyReplicas == ss.Spec.Replicas
-	failed := false
-
-	return ready, failed, nil
+	ss, ok := obj.(*kappsv1.StatefulSet)
```
Comment (deads2k) [inline pkg/template/controller/readiness.go]: you'll need all versions of statefulsets

File: pkg/template/controller/readiness.go [hunk 7]
```diff
@@ -62,105 +75,109 @@ func checkBuildConfigReadiness(oc buildclient.Interface, obj runtime.Object) (bo
...
+	route, ok := obj.(*routev1.Route)
+	if !ok {
+		return false, false, fmt.Errorf("object %T is not v1.Route", obj)
+	}
+	return len(route.Spec.Host) > 0, false, nil
 }
 
 // readinessCheckers maps GroupKinds to the appropriate function.  Note that in
 // some cases more than one GK maps to the same function.
 var readinessCheckers = map[schema.GroupKind]func(runtime.Object) (bool, bool, error){
```
Comment (deads2k) [inline pkg/template/controller/readiness.go]: this needs to be keyed by gvk

File: pkg/template/controller/templateinstance_controller.go (hunk 2 above)
Comment (deads2k) [reply pkg/template/controller/templateinstance_controller.go]: > mfojtik 5 hours ago  Member
> @deads2k candidates for openshift/api (will send a follow up with bump)

yes, move them there.

**Comment by deads2k** [comment]
@mfojtik I think we need to eat the cost of duplication for all versions and then we'll never have to worry about it again.

File: pkg/template/controller/readiness.go (hunk 5 above)
Comment (mfojtik) [reply pkg/template/controller/readiness.go]:
```
batch/v1/types.go
28:type Job struct {
```

It seems it only exists in batchv1... @soltysh can you confirm?

**Comment by mfojtik** [comment]
@deads2k all flavors for upstream workloads added, I tried to minimize the duplication and I think it is not that terrible at the end.

File: pkg/template/controller/templateinstance_controller.go (hunk 2 above)
Comment (mfojtik) [reply pkg/template/controller/templateinstance_controller.go]: @deads2k https://github.com/openshift/api/pull/87 (can replace in follow up or just bump the api in this PR after you merge it)

File: pkg/template/controller/readiness.go (hunk 5 above)
Comment (soltysh) [reply pkg/template/controller/readiness.go]: Yup, that is correct.

**Pull request merged**